src/main/java/com/openshift/jenkins/plugins/pipeline/OcWatch.java

```diff
@@ -95,132 +93,90 @@ public boolean takesImplicitBlockArgument() {
+                                        }
+                                        listener.getLogger().println("watch closure returned " + o);
+                                        return false; // don't interrupt `oc`
+                                    } catch (groovy.json.JsonException ex) {
+                                        // FIXME: We've seen instances where if the watch closer subsequently calls oc and it processes output
+                                        // before oc has finished updating it we can get json formatting
+                                        // exceptions processing the output ... we can retry here
+                                        if (jsonRetries[0]++ >= 5) {
+                                            getContext().onFailure(ex);
+                                            throw ex; // give up retrying
+                                        }
                                         listener.getLogger().println("watch closer got json formatting exception, trying again");
-                                        jsonRetries++;
-                                    } else {
+                                    } catch (InterruptedException tie) { // timeout{} block interrupted us
+                                        listener.getLogger().println("\nwatch closure interrupted (timeout?)");
+                                        getContext().onFailure(tie);
+                                        throw tie;
+                                    } catch (Throwable t) {
                                         String exceptionMsgs = t.getMessage();
                                         if (t.getCause() != null) {
                                             exceptionMsgs = exceptionMsgs + "; " + t.getCause().getMessage();
                                         }
                                         listener.getLogger().println(String.format("\nwatch closure threw an exception: \"%s\".\n", exceptionMsgs));
                                         getContext().onFailure(t);
-                                        return null;
+                                        throw new IOException(t);
                                     }
-                                } else {
-                                    String exceptionMsgs = t.getMessage();
-                                    if (t.getCause() != null) {
-                                        exceptionMsgs = exceptionMsgs + "; " + t.getCause().getMessage();
-                                    }
-                                    listener.getLogger().println(String.format("\nwatch closure threw an exception: \"%s\".\n", exceptionMsgs));
-                                    getContext().onFailure(t);
-                                    return null;
                                 }
-                            }
-
-                            if (reCheckSleep < 10000) { // Gradually check less
```

Comment:
why did you remove the progessive sleep?

during initial creation of this plugin, concerns / issues arose with hammering jenkins and the openshift master with `oc` invocations

I'll listen to your rationale, but unless it is something really unexpected, you need to add this back

Reply:
If `oc get <something> --watch` works as expected, it should print a new line every time the watch object is changed, then the closure is called because BufferedReader receives a new line.
It doesn't seem to have the necessity of rechecking the length of output, unless `oc` is buggy to print `\n`.

Reply:
oc get ... --watch is not guaranteed to stay up forever, much less for the duration of what you are watching, until it reaches whatever status / condition the closure is expecting.

You need to allow for the fact that we may have to call it multiple times.

Independent of what your reader's are doing.

Add the sleep back in.

Reply:
@gabemontero `reCheckSleep` is used to recheck the length of output, not to restart `oc get ... --watch`.

 Restarting `oc get ... --watch` is already in the loop of `while (!watchSuccess[0]);`. It restarts `oc get .. --watch` until `watchSuccess[0]` becomes `true` or an exception is caught.

Reply:
I understand that, but if it exits rapidly, you end up spinning / hammering jenkins/openshift server.

Add the progressive sleep back in

Reply:
Do you mean use `reCheckSleep` to control the frequency of restarting `oc get ... --watch` rather than checking the output change? If so, I would like to add it back.